### PR TITLE
refactor: Remote pkg & sync strategy label

### PR DESCRIPTION
**Pull request by lindnerby**
**Description**

Changes proposed in this pull request:

- Remove client caching from context and replace skr context retrieval with explicit interface
- Remove need for sync strategy annotation
- Remove global var localClientFunc  and function implementations to be replaced with context related interface implementations
- Move remote pkg to internal

**Related issue(s)**
* https://github.com/kyma-project/lifecycle-manager/pull/1426

### CODE REVIEW

File: internal/controller/kyma/controller.go [hunk 1]
```diff
@@ -115,62 +114,45 @@ func (r *KymaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
...
-
-	if !kyma.DeletionTimestamp.IsZero() && errors.Is(err, remote.ErrAccessSecretNotFound) {
-		logger.Info("access secret not found for kyma, assuming already deleted cluster")
-		r.Metrics.CleanupMetrics(kyma.Name)
-		r.removeAllFinalizers(kyma)
-
-		if err := r.updateKyma(ctx, kyma); err != nil {
-			r.Metrics.RecordRequeueReason(metrics.KymaUnderDeletionAndAccessSecretNotFound, queue.UnexpectedRequeue)
-			return ctrl.Result{}, err
+	if r.SyncKymaEnabled(kyma) {
```
Comment (ruanxin) [inline internal/controller/kyma/controller.go:117]: can we create an issue to introduce another controller for the separation of single/dual kyma reconciliation?

Comment (lindnerby) [reply internal/controller/kyma/controller.go:117]: Sure, created it [here](https://github.com/kyma-project/lifecycle-manager/issues/1525). Still needs to be fully refined.

File: internal/controller/kyma/controller.go [hunk 2]
```diff
@@ -115,62 +114,45 @@ func (r *KymaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
...
 	}
 
 	return r.reconcile(ctx, kyma)
 }
 
-func (r *KymaReconciler) deleteRemoteClientCache(ctx context.Context, kyma *v1beta2.Kyma) {
+func (r *Reconciler) deleteRemoteClientCache(ctx context.Context, kyma *v1beta2.Kyma) {
 	logger := logf.FromContext(ctx)
 	logger.Info("connection refused, assuming connection is invalid and resetting cache-entry for kyma")
 	r.RemoteClientCache.Delete(client.ObjectKeyFromObject(kyma))
```
Comment (ruanxin) [inline internal/controller/kyma/controller.go:152]: you can also use `kyma.GetNamespacedName()`, right?

File: internal/remote/client_factory.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/lifecycle-manager/api/shared"
+)
+
+type SkrContextFactory interface {
+	Get(kyma types.NamespacedName) (*SkrContext, error)
+	Init(ctx context.Context, kyma types.NamespacedName) error
```
Comment (ruanxin) [inline internal/remote/client_factory.go:22]: I think we can also introduce `DeleteCache(kyma types.NamespacedName)` to this interface, and then we can get rid of that `RemoteClientCache` in `KymaReconciler struct`

File: internal/remote/client_factory.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	restConfig.Burst = k.kcpClient.Config().Burst
+
+	clnt, err := client.New(restConfig, client.Options{Scheme: k.kcpClient.Scheme()})
+	if err != nil {
+		return fmt.Errorf("failed to create lookup client: %w", err)
+	}
+
+	skrClient := NewClientWithConfig(clnt, restConfig)
+
+	namespace := &apicorev1.Namespace{
```
Comment (ruanxin) [inline internal/remote/client_factory.go:75]: this whole "ensures remote namespace exists" logic can be extracted to the main reconciler logic, then you don't need to implement such logic in `DualClusterFactory`, `SingleClusterFactory` for the integration test. It's not a logic of this init function should care about.

File: internal/controller/kyma/controller.go (hunk 2 above)
Comment (ruanxin) [reply internal/controller/kyma/controller.go:152]: check this comment for further improvement https://github.com/kyma-project/lifecycle-manager/pull/1426/files#r1594129327

File: internal/remote/client_factory.go (hunk 4 above)
Comment (ruanxin) [reply internal/remote/client_factory.go:75]: something like this in kyma reconciler
```
if r.SyncKymaEnabled(kyma) {
    err := r.SkrContextFactory.Init(ctx, kyma.GetNamespacedName())
    skrContext, err := r.SkrContextFactory.Get(kyma.GetNamespacedName())
    if err != nil {
	    return ctrl.Result{}, err
    }
    if err := r.ensureRemoteNamespaceExits(ctx, skrContext); err != nil {
	    return ctrl.Result{}, err
    }
...
```

File: internal/controller/kyma/controller.go [hunk 5]
```diff
@@ -115,41 +114,71 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-			r.removeAllFinalizers(kyma)
-
-			if err := r.updateKyma(ctx, kyma); err != nil {
-				r.Metrics.RecordRequeueReason(metrics.KymaUnderDeletionAndAccessSecretNotFound, queue.UnexpectedRequeue)
-				return ctrl.Result{}, err
-			}
-			r.Metrics.RecordRequeueReason(metrics.KymaUnderDeletionAndAccessSecretNotFound, queue.IntendedRequeue)
-			return ctrl.Result{Requeue: true}, nil
+		result, err := r.handleRemoteInitialization(ctx, kyma)
+		if err != nil {
```
Comment (ruanxin) [inline internal/controller/kyma/controller.go:118]: you didn't catch this case `return ctrl.Result{Requeue: true}, nil`

Comment (ruanxin) [reply internal/controller/kyma/controller.go:118]: I mean [here](https://github.com/kyma-project/lifecycle-manager/pull/1426/files/834af8e799d6a3cf317c9898980788141aa0a8a1..b1dad117f6ef192357ecc852d7b3a42026af0671#diff-4d414901850a616d4e06a120615896904639aef683f0a4d8860abf65c76179d4R139)

**Review by ruanxin — APPROVED** [review]

**Review by mmitoraj — APPROVED** [review]

**Pull request merged**